cloudferrylib/os/identity/keystone.py

```diff
@@ -474,7 +491,79 @@ def _get_user_tenants_roles(self, tenant_list=None, user_list=None):
             tenant_list = []
         if user_list is None:
             user_list = []
-        user_tenants_roles = {}
+        if not self.config.identity.optimize_user_role_fetch:
+            user_tenants_roles = self._get_user_tenants_roles_by_api(tenant_list,
+                                                                     user_list)
+        else:
+            user_tenants_roles = self._get_user_tenants_roles_by_db(tenant_list,
+                                                                    user_list)
+        return user_tenants_roles
+
+    def _get_roles_sql_request(self):
+        res = []
+        try:
+            is_project_metadata = self.mysql_connector.execute(
+                "SHOW TABLES LIKE 'user_project_metadata'").rowcount
+            if is_project_metadata: #for grizzly case
+                return self.mysql_connector.execute(
+                    "SELECT * FROM user_project_metadata")
+            is_assignment = self.mysql_connector.execute(
+                "SHOW TABLES LIKE 'assignment'").rowcount
+            if is_assignment: #for icehouse case
+                res_raw = self.mysql_connector.execute(
+                    "SELECT * FROM assignment")
+                res_tmp = {}
+                for r in res_raw:
+                    if (r[1], r[2]) not in res_tmp:
+                        res_tmp[(r[1], r[2])] = {'roles': []}
+                    res_tmp[(r[1], r[2])]['roles'].append(r[3])
+                for k, v in res_tmp.iteritems():
+                    res.append((k[0], k[1], str(v)))
+        except ProgrammingError as e:
+            LOG.warn(e.message)
+        return res
+
+    def _get_user_roles_cached(self):
+        all_roles = {}
+        if self.config.identity.optimize_user_role_fetch:
+            res = self._get_roles_sql_request()
+            for row in res:
+                user_id = row[0]
+                tenant_id = row[1]
+                roles_ids = ast.literal_eval(row[2])['roles']
+                all_roles[user_id] = {} if not user_id in all_roles else all_roles[user_id]
+                all_roles[user_id][tenant_id] = [] \
+                    if not tenant_id in all_roles[user_id] else all_roles[user_id][tenant_id]
```

Comment:
Are you sure you've not violated pep8 here? 
Limit all lines to a maximum of 79 characters please (https://www.python.org/dev/peps/pep-0008/#maximum-line-length). Just set it in your IDE as a default)